pkg/sensors/accelerator/gpu/source/gpu_dcgm.go

```diff
@@ -0,0 +1,372 @@
+	d.devices = make(map[string]interface{})
+	d.entities = make(map[string]dcgm.GroupEntityPair)
+
+	cleanup, err := dcgm.Init(dcgm.Embedded)
+	if err != nil {
+		if cleanup != nil {
+			cleanup()
+		}
+		return fmt.Errorf("not able to connect to DCGM: %s", err)
+	}
+	d.cleanup = cleanup
+	dcgm.FieldsInit()
+
+	if err := d.initNVML(); err != nil {
+		d.Shutdown()
+		return err
+	}
+
+	if err := d.createDeviceGroup(); err != nil {
+		d.Shutdown()
+		return err
+	}
+
+	if err := d.addDevicesToGroup(); err != nil {
+		d.Shutdown()
+		return err
+	}
+
+	if err := d.createFieldGroup(); err != nil {
+		d.Shutdown()
+		return err
+	}
+
+	if err := d.setupWatcher(); err != nil {
+		d.Shutdown()
+		return err
+	}
+	klog.Infof("DCGM initialized successfully")
+	d.collectionSupported = true
+	return nil
+}
+
+func (d *GPUDcgm) IsGPUCollectionSupported() bool {
+	return d.collectionSupported
+}
+
+func (d *GPUDcgm) SetGPUCollectionSupported(supported bool) {
+	d.collectionSupported = supported
+}
+
+func (d *GPUDcgm) Shutdown() bool {
+	nvml.Shutdown()
+	dcgm.FieldsTerm()
+	if d.deviceGroupName != "" {
+		dcgm.DestroyGroup(d.deviceGroupHandle)
+	}
+	if d.fieldGroupName != "" {
+		dcgm.FieldGroupDestroy(d.fieldGroupHandle)
+	}
+	if d.cleanup != nil {
+		d.cleanup()
+	}
+	d.collectionSupported = false
+	return true
+}
+
+func (d *GPUDcgm) GetAbsEnergyFromGPU() []uint32 {
+	gpuEnergy := []uint32{}
+	for _, device := range d.devices {
+		power, ret := device.(nvml.Device).GetPowerUsage()
+		if ret != nvml.SUCCESS {
+			klog.V(2).Infof("failed to get power usage on device %v: %v\n", device, nvml.ErrorString(ret))
+			continue
+		}
+		// since Kepler collects metrics at intervals of SamplePeriodSec, which is greater than 1 second, it is
+		// necessary to calculate the energy consumption for the entire waiting period
+		energy := uint32(uint64(power) * config.SamplePeriodSec)
+		gpuEnergy = append(gpuEnergy, energy)
+	}
+	return gpuEnergy
+}
+
+func (d *GPUDcgm) GetGpus() map[string]interface{} {
+	return d.devices
+}
+
+func (d *GPUDcgm) GetProcessResourceUtilizationPerDevice(device interface{}, deviceName string, since time.Duration) (map[uint32]ProcessUtilizationSample, error) {
+	processAcceleratorMetrics := map[uint32]ProcessUtilizationSample{}
+
+	if device == nil { // this is a MIG device, it is already tracked in the parent device
+		return processAcceleratorMetrics, nil
+	}
+
+	var vals, miVals []dcgm.FieldValue_v1
+	var err error
+
+	klog.V(debugLevel).Infof("Device %v\n", deviceName)
+
+	deviceIndex, strErr := strconv.Atoi(deviceName)
+	if strErr != nil {
+		klog.V(debugLevel).Infof("failed to convert %q to an integer: %v", deviceName, strErr)
+		return processAcceleratorMetrics, strErr
+	}
+	vals, err = dcgm.GetLatestValuesForFields(uint(deviceIndex), deviceFields)
+	if err != nil {
+		klog.V(debugLevel).Infof("failed to get latest values for fields: %v", err)
+		return processAcceleratorMetrics, err
+	}
+	gpuSMActive := uint32(0)
```

Comment:
Same here, this is tensor utilization, not SM.
It might be better to just use `utilization`.